File: src/scripts/get_apps.py
# -*- coding: utf-8 -*-
# /usr/bin/python (macOS system python)
import json
import notifind_db

def get_apps_json(search_identifier):
  connection = notifind_db.connect()
  cursor = connection.cursor()
  apps = cursor.execute(
    """
      SELECT
        app.id,
        app.identifier,
        COUNT(record.app_id) as notification_count
      FROM app
      LEFT OUTER JOIN record
        ON app.id=record.app_id
      WHERE app.identifier LIKE ?
      GROUP BY app.id
      ORDER BY notification_count DESC;
    """,
    [
      '%' + search_identifier + '%'
    ]
  )

  res_j = []
  for row in apps:
    app_dic = {}
    app_dic['id'] = row[0]
    app_dic['identifier'] = row[1]
    app_dic['notification_count'] = row[2]
    res_j.append(app_dic)
  return res_j

def call(search_identifier):
  d = get_apps_json(search_identifier)

  json_string = json.dumps(d, indent=2, ensure_ascii=False)
  print json_string

File: src/scripts/get_apps.ts
import { execSync } from "child_process";
import { distPath } from "@src/utils/paths";
import "@src/scripts/notifind_db.sqlite3";
import "@src/scripts/notifind_db.py";
import "@src/scripts/get_apps.py";

type AppJSON = {
  id: number;
  identifier: string;
  notification_count: number;
};

export type App = {
  id: number;
  identifier: string;
  notificationCount: number;
};

export type Apps = App[];

export const getApps = (searchWord: string): Apps => {
  const jsonString = execSync(
    `(cd ${distPath} && PYTHONIOENCODING=utf-8 /usr/bin/python -c "import get_apps; get_apps.call('${searchWord}')")`
  ).toString();

  return JSON.parse(jsonString).map((json: AppJSON) => ({
    id: json.id,
    identifier: json.identifier,
    notificationCount: json.notification_count
  }));
};
